src/charm.py

```diff
@@ -1,18 +1,31 @@
 #!/usr/bin/env python3
+
 # Copyright 2021 Canonical
 # See LICENSE file for licensing details.
 
+"""Charm for creating and managing GitHub self-hosted runner instances."""
+
 import logging
-import subprocess
 import urllib.error
-from pathlib import Path
-
-from jinja2 import Environment, FileSystemLoader
-from ops.charm import CharmBase
+from typing import TYPE_CHECKING, Dict, Optional
+
+from ops.charm import (
+    ActionEvent,
+    CharmBase,
+    ConfigChangedEvent,
+    InstallEvent,
+    StopEvent,
+    UpgradeCharmEvent,
+)
 from ops.framework import EventBase, StoredState
 from ops.main import main
 from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
-from runner import RunnerManager, RunnerError, VMResources
+
+from event_timer import EventTimer, TimerDisableError, TimerEnableError
+from runner import RunnerError, RunnerManager, VMResources
+
+if TYPE_CHECKING:
+    from ops.model import JsonObject
```

Comment:
Out of curiosity, what does this do?

Reply:
The `ActionEvent.set_result` has a input type of `_SerializedData` which is defined as `Dict[str, JsonObject]`.
So we need to import `JsonObject` for a function output.

The problem here is the definition of `JsonObject` it is guarded with `if TYPE_CHECKING:`. So the definition of `JsonObject` only exists when `TYPE_CHECKING` is true, which is controlled by Python.

`if TYPE_CHECKING:` in our code is needed to prevent import `JsonObject` when it is not defined.

In practical term, when running mypy, `TYPE_CHECKING` will equal to true, and `JsonObject` will be imported to satisfy type checking. When running the charm, `TYPE_CHECKING` will equal to false, and a non-existence class will not be import. Removing the `if TYPE_CHECKING` will cause import error when running the charm.

Side note: I have tried setting the type hint of a function return to `Dict[str, Dict[str, int]]`, but mypy does not realize that type can satisfy `Dict[str, JsonObject]`.